modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Decoder.scala

```diff
@@ -125,64 +217,85 @@ trait StandardDecoderInstances { this: PrimitiveDecoderInstances =>
-        Left(DecodingFailure(errMsg, c.history))
-      case None =>
-        Left(DecodingFailure("Unable to decode Vector, expected cursor to be centered on a single Node", c.history))
-    }
-  }
-
-  implicit final def graphDecodeList[A](implicit A: Decoder[A]): Decoder[List[A]] = Decoder.instance { c =>
-    // TODO: make this stack safe
-    c.narrow.focus match {
-      case Some(IriNode(rdf.nil)) => Right(Nil)
-      case Some(_: IriOrBNode) =>
-        for {
-          head <- A(c.down(rdf.first))
-          tail <- graphDecodeList[A].apply(c.down(rdf.rest))
-        } yield head :: tail
-      case Some(_: Literal) =>
-        val errMsg =
-          "Unable to decode value to List, expected either rdf:nil or node with rdf:first and rdf:rest, but found literal."
-        Left(DecodingFailure(errMsg, c.history))
-      case None =>
-        Left(DecodingFailure("Unable to decode List, expected cursor to be centered on a single Node", c.history))
+  final def graphDecodeSeqFromList[C[_], A](f: Factory[A, C[A]])(implicit A: Decoder[A]): Decoder[C[A]] = {
+    import scala.collection.mutable
+    @tailrec
+    def inner(c: Cursor, acc: Either[DecodingError, mutable.Builder[A, C[A]]]): Either[DecodingError, C[A]] =
+      acc match {
+        case l @ Left(_) => l.asInstanceOf[Either[DecodingError, C[A]]]
+        case Right(builder) =>
+          if (c.narrow.as[AbsoluteIri] == Right(rdf.nil)) Right(builder.result())
+          else {
+            val first = c.down(rdf.first).as[A]
+            val rest  = c.down(rdf.rest)
+            inner(rest, first.map(a => builder.addOne(a)))
+          }
+      }
+
+    Decoder.instance { c =>
+      inner(c, Right(f.newBuilder))
     }
   }
 
+  implicit final def graphDecodeVector[A](implicit A: Decoder[A]): Decoder[Vector[A]] =
+    graphDecodeSeqFromList(Vector)
+
+  implicit final def graphDecodeList[A](implicit A: Decoder[A]): Decoder[List[A]] =
+    graphDecodeSeqFromList(List)
+
+  implicit final def graphDecodeArray[A: ClassTag](implicit A: Decoder[A]): Decoder[Array[A]] =
+    graphDecodeSeqFromList(Array)
+
   implicit final def graphDecodeOption[A](implicit A: Decoder[A]): Decoder[Option[A]] = Decoder.instance { c =>
     c.narrow.focus match {
       case Some(_) => A(c).map(Some.apply)
       case None    => Right(None)
     }
   }
+
+  implicit final def graphDecodeSome[A](implicit A: Decoder[A]): Decoder[Some[A]] =
+    A.map(a => Some(a))
+
+  implicit final val graphDecodeNone: Decoder[None.type] = Decoder.instance { c =>
+    c.narrow.focus match {
+      case None    => Right(None)
+      case Some(_) => Left(DecodingError("Unable to decode as None, cursor selection matches an element", c.history))
+    }
+  }
+
+  implicit final def graphDecodeEither[A, B](implicit A: Decoder[A], B: Decoder[B]): Decoder[Either[A, B]] =
+    Decoder.instance { c =>
+      A(c) match {
+        case Left(_)  => B(c).map(b => Right(b))
```

Comment:
I don't think we need here an explicit b. Why not `Right.apply` or `Right(_)` since it is obvious.

Reply:
using `b` make the expression a little more readable.